T1 linear pipeline

mdiazmel opened the pull request:
This pipeline prepare data for automatic classification pipeline. It performs three preprocessing steps :

1. Field bias correction with tool `n4biascorrection` (ANTS).
2. Registration in the space MNI152NLin2009cSym using a predefined template using `RegistrationSynQuick` tool (ANTS).
3. Crop T1 image based on a template using `nilearn`.

Templates are not included in source code, but they areautomatically downloaded into the `clinica/resources/mask/` folder.

pep8speaks commented:
Hello @mdiazmel! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:











* In the file [`clinica/utils/inputs.py`](https://github.com/aramis-lab/clinica/blob/1e0a18f0fe3ec26810205611c5d1e78f9bd17cdc/clinica/utils/inputs.py):

> [Line 340:13](https://github.com/aramis-lab/clinica/blob/1e0a18f0fe3ec26810205611c5d1e78f9bd17cdc/clinica/utils/inputs.py#L340): [E117](https://duckduckgo.com/?q=pep8%20E117) over-indented

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_cli.py:
```diff
@@ -0,0 +1,92 @@
+        optional = self._args.add_argument_group(PIPELINE_CATEGORIES['OPTIONAL'])
+        optional.add_argument("-rt", "--ref_template",
+                              help='Reference template for registration.')
+
```
Quick question: why the reference template needs to be provided by the user? I saw you put a downloading mechanism for this template.

Currenly, for any atlas, either it is embedded in Clinica or it embedded via neuroimaging software packages (e.g. JHU* for FSL, Neuromorphometrics & cie for CAT12).

mdiazmel replied:
> Quick question: why the reference template needs to be provided by the user? I saw you put a downloading mechanism for this template.

I just kept it as an optional argument (e.g., if user wants to change the template). I don't known if it's really useful. I will remove it in the next commit.

> Currenly, for any atlas, either it is embedded in Clinica or it embedded via neuroimaging software packages (e.g. JHU* for FSL, Neuromorphometrics & cie for CAT12).

I don't want to distribute the templates with the package (in order to preserve a light package). They made almost 50Mb. I'm trying to add this "downloading mechanism" as a method of the module `clinica/utils/inputs.py`. It can be useful in other cases.

mdiazmel commented:
Hi @alexandreroutier , this PR is ready for review. 
Normally, CI should pass. Could you take a look when you have some time? 
Thanks,

alexandreroutier commented on clinica/pipelines/t1_linear/info.json:
```diff
@@ -0,0 +1,12 @@
+            "version": ">=0.1.0"
+        }
+    ]
+}
```
Your `info.json` should look like
```JSON
{
    "id": "aramislab/t1-linear",
    "author": " Mauricio Diaz",
    "version": "0.1.0",
    "space_caps": "42M",
    "space_wd": "42M",
    "dependencies": [
        {
            "type": "software",
            "name": "ants",
            "version": ">=2.3.1"
        }
    ]
} 
```
space_caps/space_caps were added recently: this is the size of the outputs for a single (participant, session). So replace these default values by the adequate one.

alexandreroutier commented on clinica/cmdline.py:
```diff
@@ -226,7 +227,8 @@ def execute():
         T1VolumeDartel2MNICLI(),
-        T1VolumeParcellationCLI()
+        T1VolumeParcellationCLI(),
+        T1LinearCLI()
```
It would be better to put `T1LinearCLI` between T1FreeSurferLongitudinalCLI() and DwiPreprocessingUsingPhaseDiffFieldmapCli(). When typing `clinica run`, `t1-linear` will displayed before `dwi-preprocessing` in the list of pipelines.

`pipelines` contains main pipelines sorted by modalities. 'Advanced' pipelines are displayed at the end of the list.

alexandreroutier commented on clinica/pipelines/t1_linear/README.md:
```diff
@@ -0,0 +1,70 @@
+# `t1-linear`` - <VERY_SHORT_DESCRIPTION>
```
Can be deleted.

(I'm also wondering if we should remove this from `clinica generate` :thinking: )

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_cli.py:
```diff
@@ -0,0 +1,93 @@
+                                 help='Temporary directory to store pipelines intermediate results.')
+        clinica_opt.add_argument("-np", "--n_procs",
+                                 metavar=('N'), type=int,
+                                 help='Number of cores used to run in parallel.')
```
To be replaced by:
```python
        # Clinica standard arguments (e.g. --n_procs)
        self.add_clinica_standard_arguments()
```

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_cli.py:
```diff
@@ -0,0 +1,93 @@
+            pipeline.run(plugin='MultiProc',
+                         plugin_args={'n_procs': args.n_procs})
+        else:
+            pipeline.run()
```
I think the Jinja2 was not updated when you generated the skeleton. The `run_command` should look like:

```python
    def run_command(self, args):
        """Run the pipeline with defined args."""
        from networkx import Graph
        from .t1_volume_tissue_segmentation_pipeline import T1Linear
        from clinica.utils.ux import print_end_pipeline, print_crash_files_and_exit
    
        parameters = {
            #'ref_template'        : args.ref_template or 'Reference Template'
        }
        pipeline = T1Linear(
            bids_directory=self.absolute_path(args.bids_directory),
            caps_directory=self.absolute_path(args.caps_directory),
            tsv_file=self.absolute_path(args.subjects_sessions_tsv),
            base_dir=self.absolute_path(args.working_directory),
            parameters=parameters,
            name=self.name
        )
    
        if args.n_procs:
            exec_pipeline = pipeline.run(plugin='MultiProc',
                                         plugin_args={'n_procs': args.n_procs})
        else:
            exec_pipeline = pipeline.run()
    
        if isinstance(exec_pipeline, Graph):
            print_end_pipeline(self.name, pipeline.base_dir, pipeline.base_dir_was_specified)
        else:
            print_crash_files_and_exit(args.logname, pipeline.base_dir)
```

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_pipeline.py:
```diff
@@ -0,0 +1,263 @@
+"""T1 Linear - Clinica Pipeline.
```
Missing # coding: utf-8 (this is corrected in PR #63 if you generate new pipeline.

BTW, take the opportunity to delete the comments which were automatically generated by clinica (they were designed to help the developper. If you didn't feel that way, it could be nice to have your feedback ^^).

arnaudmarcoux commented on test/nonregression/test_run_pipelines.py:
```diff
@@ -699,6 +699,48 @@ def test_run_SpatialSVM(cmdopt):
+        tsv_file=join(root, 'in', 'subjects.tsv'),
+        base_dir=join(working_dir, 'T1Linear'),
+    )
+    pipeline.build()
```
pipeline is automatically build when it is `run()` so this line is not necessary

alexandreroutier replied:
Indeed. But this is also the case for all other `test_run_Pipeline()`.

@mdiazmel : Could you take the opportunity of this PR to remove `pipeline.build()` for `test_run_pipelines.py`

mdiazmel replied:
Done!

mdiazmel commented:
@alexandreroutier: this PR is ready for a second pass. I addressed all the issues and I also added the option to produce (or not) the cropped version of the image.

alexandreroutier commented on clinica/pipelines/t1_linear/README.md:
```diff
@@ -0,0 +1,68 @@
+
+!!! tip
+    Easily access the papers cited on this page on
+    [Zotero](https://www.zotero.org/groups/1517933/aramis_clinica/items/collectionKey/2DHP3WXH).
```
File to delete.

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_cli.py:
```diff
@@ -0,0 +1,82 @@
+This file has been generated automatically by the `clinica generate template`
+command line tool. See here for more details:
+http://clinica.run/doc/InteractingWithClinica/
+"""
```
Can be deleted (not needed anymore)

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_cli.py:
```diff
@@ -0,0 +1,82 @@
+
+        parameters = {
+                'crop_image': args.crop_image
+                # 'ref_template'        : args.ref_template or 'Reference Template'
```
Can be deleted (not needed anymore).

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_pipeline.py:
```diff
@@ -0,0 +1,258 @@
+"""T1 Linear - Clinica Pipeline.
+This file has been generated automatically by the `clinica generate template`
+command line tool. See here for more details: https://gitlab.icm-institute.org/aramislab/clinica/wikis/docs/InteractingWithClinica.
+"""
```
Can be deleted (not needed anymore)

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_pipeline.py:
```diff
@@ -0,0 +1,258 @@
+        >>> }
+        >>> pipeline.base_dir = '/tmp/'
+        >>> pipeline.run()
+    """
```
Proposition:
```python
class T1Linear(cpe.Pipeline):
    """T1Linear - Affine registration of T1w images to standard space.

    This pipeline performs globally three steps:
    1) Bias correction with N4 algorithm from ANTs.
    2) Linear registration to MNI152NLin2009cSym template
    with RegistrationSynQuick from ANTs.
    3) Crop of the the background (in order to save computational power)

    Returns:
        A clinica pipeline object containing the T1Linear pipeline.
    """
```
Currently, pipeline classes do not contain pipeline args section nor example section. This can be rediscussed when a Sphinx or similar tool will be added to Clinica.

mdiazmel replied:
>> Currently, pipeline classes do not contain pipeline args section nor example section. This can be rediscussed when a Sphinx or similar tool will be added to Clinica.

That's a pity! We should start thinking about fully documented functions from now!
I accept you proposition in order to get forward with the PR.

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_pipeline.py:
```diff
@@ -0,0 +1,258 @@
+
+        check_bids_folder(self.bids_directory)
+        is_bids_dir = True
+
```
Not needed, this is automatically performed in Clinica.

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_pipeline.py:
```diff
@@ -0,0 +1,258 @@
+        path_to_mask = join(root, 'resources', 'masks')
+        self.ref_template = join(
+                path_to_mask, 'mni_icbm152_t1_tal_nlin_sym_09c.nii')
+        self.ref_crop = join(path_to_mask, 'ref_cropped_template.nii.gz')
```
Naive question: why `ref_template` and  `ref_crop` are attributes of the class and do not belong to parameters dictionary?

mdiazmel replied:
I don't have a strong argument to support this choice. The only thing that I can say is that I see the parameters dictionary as the user input values (user can change them). The templates to register/crop are already fixed by the developers.

alexandreroutier replied:
Ok :upside_down_face:

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_pipeline.py:
```diff
@@ -0,0 +1,258 @@
+                is_bids_dir,
+                False,
+                self.base_dir
+                )
```
This is automatically done in Abstract class (engine.py)

mdiazmel replied:
ok, deleted!

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_pipeline.py:
```diff
@@ -0,0 +1,258 @@
+            try:
+                fetch_file(url2, self.ref_template)
+            except IOError as err:
+                cprint('Unable to download required template (mni_icbm152) for processing:', err)
```
Why the exception is not raised?

mdiazmel replied:
How do you tested ? (You tried to generate an `IOError`?)

alexandreroutier replied:
From what I read on your, if a `IOError` is caught, only the `cprint` is displayed meaning that `self.ref_template` will not contain :thinking: 

I had no issue to run your pipeline for information :) but I don't know how a `IOError` could happen (lack of space) and was wondering how to handle this.

Anyway, this was a minor question.

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_pipeline.py:
```diff
@@ -0,0 +1,258 @@
+                    self.bids_directory,
+                    T1W_NII)
+        except ClinicaException as e:
+            err = 'Clinica faced error(s) while trying to read files in your CAPS directory.\n' + str(e)
```
```suggestion
            err = 'Clinica faced error(s) while trying to read files in your BIDS directory.\n' + str(e)
```

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_utils.py:
```diff
@@ -0,0 +1,120 @@
+    cmd = ' '.join(commands)
+    os.system(cmd)
+
+    return output_img
```
Can be deleted (not used anymore)

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_pipeline.py:
```diff
@@ -0,0 +1,258 @@
+                    input_names=['bids_or_caps_filename'],
+                    output_names=['container'],
+                    function=container_from_filename),
+                name='ContainerPath')
```
The get_ids/containter_path nodes are usually set in `build_output_node` method. See e.g. [`pet_volume_pipeline.py`](https://github.com/aramis-lab/clinica/blob/dev/clinica/pipelines/pet_volume/pet_volume_pipeline.py) to see its use

mdiazmel replied:
Done!

alexandreroutier commented:
Hi @mdiazmel,

I finished the code review. Except one comment regarding get_ids/containter_path nodes and some cleaning, this PR seems to be ready to merge. :)

alexandreroutier commented on test/instantiation/test_instantiate_all_pipelines.py:
```diff
@@ -349,3 +349,19 @@ def test_instantiate_T1FreeSurferLongitudinalCorrection():
+        caps_directory=join(root, 'in', 'caps'),
+        tsv_file=join(root, 'in', 'subjects.tsv'),
+    )
+    pipeline.parameters['nproc'] = 4
```
Can be removed:
```suggestion
```

mdiazmel commented:
Hi @alexandreroutier, I addressed your comments and answered your questions (some of them were answered with other questions ;)...). I let you decide if we continue the discussion or otherwise merge.

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_pipeline.py:
```diff
@@ -0,0 +1,223 @@
+# -*- coding: utf-8 -*-
+import clinica.pipelines.engine as cpe
+
+
```
When I run your pipeline, I noticed that hases were not generated in working directory:
```text
$ pwd
/home/alexandre.routier/Temp/LightDataset/WD/t1-linear
$ ls
d3.js
graph1.json
graph.json
index.html
_t1w_..home..alexandre.routier..Temp..LightDataset..BIDS_T1_LINEAR..sub-ADNI003S4081..ses-M00..anat..sub-ADNI003S4081_ses-M00_T1w.nii
_t1w_..home..alexandre.routier..Temp..LightDataset..BIDS_T1_LINEAR..sub-CAPP01001TMM3B0..ses-M00..anat..sub-CAPP01001TMM3B0_ses-M00_T1w.nii.gz
_t1w_..home..alexandre.routier..Temp..LightDataset..BIDS_T1_LINEAR..sub-CAPP01011PTC..ses-M00..anat..sub-CAPP01011PTC_ses-M00_T1w.nii
```
When using iterables, folder names can be very long and possibly generate ` OSError`. In (almost?) all pipelines, we added the following lines:

```python
# Use hash instead of parameters for iterables folder names
# Otherwise path will be too long and generate OSError
from nipype import config
cfg = dict(execution={'parameterize_dirs': False})
config.update_config(cfg)
```

This has been added to PR #63 regarding Jinja2 hoping we won't face this issue again. :)

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_pipeline.py:
```diff
@@ -0,0 +1,223 @@
+
+
+class T1Linear(cpe.Pipeline):
+    """T1 Linear SHORT DESCRIPTION.
```
I propose to use the tagline of the future Wiki page :)
```suggestion
    """T1 Linear - Affine registration of T1w images to standard space.
```

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_pipeline.py:
```diff
@@ -0,0 +1,223 @@
+        except ClinicaException as e:
+            err = 'Clinica faced error(s) while trying to read files in your BIDS directory.\n' + str(e)
+            raise ClinicaBIDSError(err)
+
```
Currenly, nothing is displayed when running the pipeline:
```text
$ clinica run t1-linear BIDS_T1_LINEAR/ CAPS_T1_LINEAR -np 2 -wd WD
[14:08:10] The t1-linear pipeline has completed. You can now delete the working directory (/home/alexandre.routier/Temp/LightDataset/WD/t1-linear).
```
The minimum information I currently demand is to display the image(s) to process and add estimation of time computation. You can simply add these lines:

```python
        if len(self.subjects):
            print_images_to_process(self.subjects, self.sessions)
            cprint('The pipeline will last approximately 42 minutes per image.') # Replace by adequate computational time.
```

These lines were added to the Jinja template (PR #63) but after you generated the skeleton.

alexandreroutier commented on clinica/pipelines/t1_linear/t1_linear_pipeline.py:
```diff
@@ -0,0 +1,223 @@
+            err = 'Clinica faced error(s) while trying to read files in your BIDS directory.\n' + str(e)
+            raise ClinicaBIDSError(err)
+
+        # Read tsv file and load inputs
```
To delete or update (you don't read TSV file anymore :upside_down_face: )

alexandreroutier commented:
Hi @mdiazmel,

I left small comments for this last review.

I have run the pipeline and noticed small issues with the CAPS:
![T1Linear](https://user-images.githubusercontent.com/24808663/75356818-bc9b3c80-58b0-11ea-90ca-ae776992ecfa.png)

- `linear_registration` needs to be removed
- In the CAPS Specification `<source_file>` is the BIDS without its extension (e.g. `sub-CLNC01_ses-M00_acq-HighRes_T1w` for BIDS file `sub-CLNC01_ses-M00_acq-HighRes_T1w.nii[.gz]`). You can see example for DWI files in the [Intro page](http://www.clinica.run/doc/CAPS/Introduction/#subject-and-group-naming). While reading the specifications, I don't see explanations about it so I will clarify it. Except `t1-freesurfer` (and `pet-surface` ? @arnaudmarcoux) CAPS files embed the BIDS files in `<source_file>`.


For you pipeline, I think you can replace:
```python
def get_data_datasink(image_id):
    substitutions_ls = [  # registration
            (image_id + '_T1w_corrected.nii.gz',
                image_id + '_corrected_T1w.nii.gz'),
            (image_id + 'Warped_cropped_intensity_norm.nii.gz',
                image_id + '_space-MNI152NLin2009cSym_res-1x1x1_intensity_norm_T1w.nii.gz'),
            (image_id + 'Warped_cropped.nii.gz',
                image_id + '_space-MNI152NLin2009cSym_desc-Crop_res-1x1x1_T1w.nii.gz'),
            (image_id + '0GenericAffine.mat',
                image_id + '_space-MNI152NLin2009cSym_res-1x1x1_affine.mat'),
            (image_id + 'Warped_cropped.pt',
                image_id + '_space-MNI152NLin2009cSym_res-1x1x1_T1w.pt'),
            (image_id + 'Warped.nii.gz',
                image_id + '_space-MNI152NLin2009cSym_res-1x1x1_linear_registration_T1w.nii.gz')
            ]
    return image_id, substitutions_ls
```
by:
```python
def get_data_datasink(bids_file):

    from nipype.utils.filemanip import split_filename

    _, source_file, _ = split_filename(bids_file)

    substitutions_ls = [  # registration
            (source_file + '_T1w_corrected.nii.gz',
                source_file + '_corrected_T1w.nii.gz'),
            (source_file + 'Warped_cropped_intensity_norm.nii.gz',
                source_file + '_space-MNI152NLin2009cSym_res-1x1x1_intensity_norm_T1w.nii.gz'),
            (source_file + 'Warped_cropped.nii.gz',
                source_file + '_space-MNI152NLin2009cSym_desc-Crop_res-1x1x1_T1w.nii.gz'),
            (source_file + '0GenericAffine.mat',
                source_file + '_space-MNI152NLin2009cSym_res-1x1x1_affine.mat'),
            (source_file + 'Warped_cropped.pt',
                source_file + '_space-MNI152NLin2009cSym_res-1x1x1_T1w.pt'),
            (source_file + 'Warped.nii.gz',
                source_file + '_space-MNI152NLin2009cSym_res-1x1x1_linear_registration_T1w.nii.gz')
            ]
    return source_file, substitutions_ls
```
and adapts the nodes in your pipeline.

Once these small issues solved, we can merge your PR :)

Alex

PS : BTW, what:
```python
[...]
            (image_id + '_T1w_corrected.nii.gz',
                image_id + '_corrected_T1w.nii.gz'),
            (image_id + 'Warped_cropped_intensity_norm.nii.gz',
                image_id + '_space-MNI152NLin2009cSym_res-1x1x1_intensity_norm_T1w.nii.gz'),
[...]
            (image_id + 'Warped_cropped.pt',
                image_id + '_space-MNI152NLin2009cSym_res-1x1x1_T1w.pt'),
            ]
    return image_id, substitutions_ls
```
are they used for? I think `_intensity_norm_T1w.nii.gz` and  `_T1w.pt` are not used in this pipeline. If `_corrected_T1w.nii.gz` corresponds to bias corrected image, I would advice something like `_desc-BiasCorrected_T1w.nii.gz` or  `_desc-BC_T1w.nii.gz` instead.

mdiazmel commented:
@alexandreroutier I finished with the modifications, but it seems that there are problems with the CI data... once this is fixed we can consider to merge the PR.
